lib/stringtheory/src/lib.rs

```diff
@@ -261,78 +650,162 @@ impl fmt::Display for MetaString {
-        unsafe { std::str::from_utf8_unchecked(s) }
+        assert_eq!("hello", &*meta);
+        assert_eq!(meta.inner.get_union_type(), UnionType::Owned);
     }
-}
 
-#[cfg(test)]
-mod tests {
-    use super::*;
+    #[test]
+    fn inlined_string() {
+        let s = "hello";
+        let meta = MetaString::from(s);
+
+        assert_eq!("hello", &*meta);
+        assert_eq!(meta.inner.get_union_type(), UnionType::Inlined);
+    }
 
     #[test]
-    fn struct_layout() {
-        assert_eq!(std::mem::size_of::<MetaString>(), 32);
+    fn interned_string() {
+        let intern_str = "hello interned str!";
+
+        let interner = FixedSizeInterner::<1>::new(NonZeroUsize::new(1024).unwrap());
+        let s = interner.try_intern(intern_str).unwrap();
+        assert_eq!(intern_str, &*s);
+
+        let meta = MetaString::from(s);
+        assert_eq!(intern_str, &*meta);
+        assert_eq!(meta.inner.get_union_type(), UnionType::Interned);
     }
 
     #[test]
-    fn inlined_string() {
-        // We expect this to hold because we need `InlinedString` to be one word less (usize, 8 bytes on 64-bit
-        // platforms) than `MetaString` to ensure it isn't causing `MetaString` to be larger than it should be, but also
-        // that it's maximizing the potential inlining capacity.
-        assert_eq!(
-            std::mem::size_of::<InlinedString>(),
-            std::mem::size_of::<MetaString>() - std::mem::size_of::<usize>()
-        );
+    fn empty_string_interned() {
+        let intern_str = "";
+
+        let interner = FixedSizeInterner::<1>::new(NonZeroUsize::new(1024).unwrap());
+        let s = interner.try_intern(intern_str).unwrap();
+        assert_eq!(intern_str, &*s);
+
+        let meta = MetaString::from(s);
+        assert_eq!(intern_str, &*meta);
+        assert_eq!(meta.inner.get_union_type(), UnionType::Empty);
+    }
+
+    #[test]
+    fn empty_string_static() {
+        let s = "";
+
+        let meta = MetaString::from_static(s);
+        assert_eq!(s, &*meta);
+        assert_eq!(meta.inner.get_union_type(), UnionType::Empty);
+    }
+
+    #[test]
+    fn empty_string_inlined() {
+        let s = "";
+
+        let meta = MetaString::try_inline(s).expect("empty string definitely 'fits'");
+        assert_eq!(s, &*meta);
+        assert_eq!(meta.inner.get_union_type(), UnionType::Empty);
     }
 
     #[test]
-    fn try_inline() {
-        // String is < 31 bytes (13 bytes), so it should be inline-able:
-        let s = "hello, world!";
-        assert_eq!(s.len(), 13);
+    fn empty_string_owned() {
+        // When a string has capacity, we don't care if it's actually empty or not, because we want to preserve the
+        // allocation... so our string here is empty but _does_ have capacity.
+        let s = String::with_capacity(4);
+
+        let meta = MetaString::from(s);
+        assert_eq!("", &*meta);
+        assert_eq!(meta.inner.get_union_type(), UnionType::Owned);
```

Comment:
None of the tests call into_owned? It'd be particularly interesting here, since you could verify that you get a string with 4 capacity at the end.

Reply:
Ooh, that's a good point, in order to test the invariant that owned strings going in are the same coming out.

Reply:
Addressed in c483845.

Reply:
Just curious, is the minimum allocation 32?